### PR TITLE
Update mcore parallelism initialization in nemo2

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Update to use encoder_tp_size and encoder_pp_size in MCore API. This enables more flexible parallelism in encoder-decoder models and VLMs.

**Collection**: [common]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/megatron_init.py [hunk 1]
```diff
@@ -95,6 +95,8 @@ def initialize_model_parallel_for_nemo(
     virtual_pipeline_model_parallel_size=None,
     pipeline_model_parallel_split_rank=None,
     context_parallel_size=1,
+    encoder_tensor_model_parallel_size=0,
+    encoder_pipeline_model_parallel_size=0,
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_init.py:99]: Should the default be `0` or `None`?

File: nemo/collections/nlp/modules/common/megatron/megatron_init.py [hunk 2]
```diff
@@ -149,12 +155,14 @@ def initialize_model_parallel_for_nemo(
     set_expert_model_parallel_world_size(app_state.expert_model_parallel_size)
     set_expert_model_parallel_rank(app_state.expert_model_parallel_rank)
 
+    set_pipeline_model_parallel_world_size(
+        app_state.pipeline_model_parallel_size + app_state.encoder_pipeline_model_parallel_size
+    )
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_init.py:160]: Okay so `pipeline_model_parallel_size` is really the decoder pipeline size now?

**Review by ericharper — COMMENTED** [review]
Can we run the nightly CI for this PR before merging?

File: nemo/collections/nlp/modules/common/megatron/megatron_init.py (hunk 1 above)
Comment (yaoyu-33) [reply nemo/collections/nlp/modules/common/megatron/megatron_init.py:99]: 0 following mcore here https://github.com/NVIDIA/Megatron-LM/blob/98abe37866bba8aa0eee246fdac5163f5c8bcff7/megatron/core/parallel_state.py#L354

File: nemo/collections/nlp/modules/common/megatron/megatron_init.py (hunk 2 above)
Comment (yaoyu-33) [reply nemo/collections/nlp/modules/common/megatron/megatron_init.py:160]: yep

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Review by huvunvidia — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**